<a href="https://colab.research.google.com/github/martinacaramaschi/TPT-PE-thematic-analysis/blob/main/02TPT_PE_articles_cleaning_v03ipynb.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# TPT and PE articles cleaning
*Update: 18th November 2024*

*   Notebook for data cleaning

# Importing datasets

In [ ]:
# importing PE journal documents
import pickle

directory_name = '/content/drive/MyDrive/Colab Notebooks/TPT_PE_review/'
document_name = 'physics_education_metadata_with_fulltext.pkl'

with open(directory_name + document_name, 'rb') as f:
    pe_data = pickle.load(f)

In [ ]:
# importing TPT journal documents
import pickle

directory_name = '/content/drive/MyDrive/Colab Notebooks/TPT_PE_review/'
document_name = 'TPT_metadata_cleaned_fulltext.pkl'

with open(directory_name + document_name, 'rb') as f:
    tpt_data = pickle.load(f)

Convert the elements of "years" in numbers

In [ ]:
import pandas as pd
# Convert the 'publication_year' column to numeric
pe_data['publication_year'] = pd.to_numeric(pe_data['publication_year'], errors='coerce')

Looking at initial rows of pe_data

In [ ]:
pe_data.head(3)

title         authors publication_year  \
0     The polar mesosphere  Morris, Murphy             2008   
1  Matter for illumination    Joan L Hawes             1981   
2                  Reviews                             2007   

                          doi volume issue fpage lpage     pdf_filename  \
0  10.1088/0031-9120/43/4/003     43     4   366   374    pe8_4_003.pdf   
1  10.1088/0031-9120/16/3/310     16     3   178   185  pev16i3p178.pdf   
2  10.1088/0031-9120/42/4/M04     42     4   414   423   pe_7_4_m04.pdf   

                                        zip_filename  \
0  1_0031-9120_43_4_366_10__1088_0031-9120_43_4_0...   
1  1_0031-9120_16_3_178_10__1088_0031-9120_16_3_3...   
2  1_0031-9120_42_4_414_10__1088_0031-9120_42_4_M...   

                                            fulltext  
0  www.iop.org/journals/physed \nThe polar mesosp...  
1  Phys Educ Val 16.1981 Printed In GreatBrltaln ...  
2  BOOK : UNCERTAIN SCIENCE …U NCERTAIN WORLD \nT...

In [ ]:
tpt_data.head(3)

index       filename  year  \
0      0  1.2350554.pdf  1963   
1      2  1.2350556.pdf  1963   
2      3  1.2350557.pdf  1963   

                                               title  author_list volume  \
0    New Unified Scale for Atomic Masses and Weights    [Nier AO]      1   
1  International Agreements on the Measurement of...  [Wichers E]      1   
2                          Weight and Weightlessness   [Sears FW]      1   

  issue  processed_len   page  page_len  ...  \
0     1          36323  11-16         6  ...   
1     1          27485  17-45        29  ...   
2     1          18790  20-23         4  ...   

                                              author  editor  reference-count  \
0  [{'given': 'Alfred O.', 'family': 'Nier', 'seq...      []                0   
1  [{'given': 'Edward', 'family': 'Wichers', 'seq...      []                0   
2  [{'given': 'Francis W.', 'family': 'Sears', 's...      []                0   

   is-referenced-by-count     issued  \
0                       1  [1963, 4]   
1                       0  [1963, 4]   
2                       3  [1963, 4]   

                                                link        doi  \
0  http://aapt.scitation.org/doi/pdf/10.1119/1.23...  1.2350554   
1  http://aapt.scitation.org/doi/pdf/10.1119/1.23...  1.2350556   
2  http://aapt.scitation.org/doi/pdf/10.1119/1.23...  1.2350557   

                                            fulltext  \
0  VOLUME I -NUMBER 1 \nT H E \nAPRIL 1963 \n\nPH...   
1  International Agreements on the Measurement \n...   
2  Weight and Weightlessness \nA definition of 11...   

                 first_n_words  \
0            NEW UNIFIED SCALE   
1  INTERNATIONAL AGREEMENTS ON   
2    WEIGHT AND WEIGHTLESSNESS   

                                    cleaned_fulltext  
0  VOLUME I -NUMBER 1 \nT H E \nAPRIL 1963 \n\nPH...  
1  International Agreements on the Measurement \n...  
2  Weight and Weightlessness \nA definition of 11...  

[3 rows x 34 columns]

# Analysing the size of specific document series


## TPT

### **Teachers' pets**

In [ ]:
# Visualise a semple of tpt_data[['title', 'author_list', 'doi', 'raw', 'fulltext']] containing "pets" in "title"

tpt_data_pets = tpt_data[tpt_data['title'].str.contains('pets', case=False)][['title','year', 'author_list', 'doi', 'raw', 'fulltext']]
tpt_data_pets.shape

(7, 6)

In [ ]:
max_year = tpt_data_pets['year'].max()
min_year = tpt_data_pets['year'].min()

print(max_year)
print(min_year)

1977
1975


### **LETTERS**

In [ ]:
tpt_data_LETTERS = tpt_data[tpt_data['title'].str.contains('LETTERS', case=True)][['title','year', 'author_list', 'doi', 'raw', 'fulltext']]
tpt_data_LETTERS.shape

(42, 6)

In [ ]:
max_year = tpt_data_LETTERS['year'].max()
min_year = tpt_data_LETTERS ['year'].min()

print(max_year)
print(min_year)

1974
1968


### **NOTES**


In [ ]:
tpt_data_NOTES = tpt_data[tpt_data['title'].str.contains('NOTES', case=True)][['title','year', 'author_list', 'doi', 'raw', 'fulltext']]

max_year = tpt_data_NOTES['year'].max()
min_year = tpt_data_NOTES ['year'].min()

print(tpt_data_NOTES.shape)
print(max_year)
print(min_year)

(292, 6)
1974
1967


### **RESEARCH FRONTIER**

In [ ]:
tpt_data_RF = tpt_data[tpt_data['title'].str.contains('RESEARCH FRONTIER', case=True)][['title','year', 'author_list', 'doi', 'raw', 'fulltext']]

max_year = tpt_data_RF['year'].max()
min_year = tpt_data_RF ['year'].min()

print(tpt_data_RF.shape)
print(max_year)
print(min_year)

(19, 6)
1974
1968


### **Microreview**

In [ ]:
tpt_data_Micro = tpt_data[tpt_data['title'].str.contains('Microreview', case=False)][['title','year', 'author_list', 'doi', 'raw', 'fulltext']]

max_year = tpt_data_Micro['year'].max()
min_year = tpt_data_Micro ['year'].min()

print(tpt_data_Micro.shape)
print(max_year)
print(min_year)

(0, 6)
nan
nan


### **Apparatus for teaching physics**

In [ ]:
tpt_data_aftp = tpt_data[tpt_data['title'].str.contains('apparatus for teaching physics', case=False)][['title','year', 'author_list', 'doi', 'raw', 'fulltext']]

max_year = tpt_data_aftp['year'].max()
min_year = tpt_data_aftp ['year'].min()

print(tpt_data_aftp.shape)
print(max_year)
print(min_year)

(189, 6)
1996
1963


### **TPT notes**

In [ ]:
tpt_data_tpt_notes = tpt_data[tpt_data['title'].str.contains('tpt notes', case=False)][['title','year', 'author_list', 'doi', 'raw', 'fulltext']]

max_year = tpt_data_tpt_notes['year'].max()
min_year = tpt_data_tpt_notes ['year'].min()

print(tpt_data_tpt_notes.shape)
print(max_year)
print(min_year)

(42, 6)
1970
1968


## PE

### **END RESULTS**

In [ ]:
pe_data_ENDR = pe_data[pe_data['title'].str.contains('ENDR', case=False)][['title','publication_year', 'authors', 'doi', 'fulltext']]

max_year = pe_data_ENDR['publication_year'].max()
min_year = pe_data_ENDR ['publication_year'].min()

print(pe_data_ENDR.shape)
print(max_year)
print(min_year)

(1, 5)
1979
1979


### **Letters**

In [ ]:
pe_data_Letters = pe_data[pe_data['title'].str.contains('Letters', case=True)][['title','publication_year', 'authors', 'doi', 'fulltext']]

max_year = pe_data_Letters['publication_year'].max()
min_year = pe_data_Letters ['publication_year'].min()

print(pe_data_Letters.shape)
print(max_year)
print(min_year)

(49, 5)
2014
1998


### **Reviews**

In [ ]:
pe_data_Reviews = pe_data[pe_data['title'].str.contains('Reviews', case=True)][['title','publication_year', 'authors', 'doi', 'fulltext']]

max_year = pe_data_Reviews['publication_year'].max()
min_year = pe_data_Reviews ['publication_year'].min()

print(pe_data_Reviews.shape)
print(max_year)
print(min_year)

(120, 5)
2022
2002


### **Editor's note**

In [ ]:
pe_data_EDnotes = pe_data[pe_data['title'].str.contains("Editor's note", case=True)][['title','publication_year', 'authors', 'doi', 'fulltext']]

max_year = pe_data_EDnotes['publication_year'].max()
min_year = pe_data_EDnotes ['publication_year'].min()

print(pe_data_EDnotes.shape)
print(max_year)
print(min_year)

(1, 5)
1998
1998


**Elements that will be removed from PE**

In [ ]:
pe_data_removed = pe_data[pe_data['title'].str.contains('book review', case=False)][['title','publication_year', 'authors', 'doi', 'fulltext']]
print(pe_data_removed.shape)

In [ ]:
pe_data_removing = pe_data[pe_data['title'].str.contains('News', case=True)][['title','publication_year', 'authors', 'doi', 'fulltext']]
pe_data_removing.sample(5)

title publication_year authors                         doi  \
130   News             2014          10.1088/0031-9120/49/3/261   
3495  News             2014            10.1088/0031-9120/49/1/1   
3962  News             2008          10.1088/0031-9120/43/6/M01   
7569  News             2015            10.1088/0031-9120/50/1/1   
5040  News             2005          10.1088/0031-9120/40/4/M01   

                                               fulltext  
130   UK public libraries offer walk-in access to re...  
3495  The fabulousness of physics in 2013 \n\nThe fr...  
3962  YOUR NEWS WANTED \nThe news section gives upda...  
7569  NEWS \niopscience.org/ped \nNurturing the rich...  
5040  YOUR NEWS WANTED \nThe news section gives upda...

# PE and TPT cleaning

## **Removing specific series using a keyword in the title**

In [ ]:
# PE
import pandas as pd

# List of keywords or phrases you want to remove from titles
keywords_to_remove1 = ['News', 'SIGNING OFF', 'PEOPLE', 'People']
keywords_to_remove2 = ['Reply to comment', 'Erratum', 'book review']



# Create a condition that checks if any of the keywords are in the 'title' column
condition1 = pe_data['title'].str.contains('|'.join(keywords_to_remove1), case=True, na=False)
condition2 = pe_data['title'].str.contains('|'.join(keywords_to_remove2), case=False, na=False)

# Filter the dataset by inverting the condition to keep rows that don't match the keywords
pe_data_cleaned = pe_data[~condition1].copy()
pe_data_cleaned = pe_data_cleaned[~condition2]

# Display the cleaned dataset
print(pe_data_cleaned.shape)


In [ ]:
# TPT
import pandas as pd

# List of keywords or phrases you want to remove from titles
keywords_to_remove1 = ['Correction', 'BOOKS FOR YOUNG PEOPLE']
keywords_to_remove2 = ['century textbook illustrations']



# Create a condition that checks if any of the keywords are in the 'title' column
condition1 = tpt_data['title'].str.contains('|'.join(keywords_to_remove1), case=True, na=False)
condition2 = tpt_data['title'].str.contains('|'.join(keywords_to_remove2), case=False, na=False)

# Filter the dataset by inverting the condition to keep rows that don't match the keywords
tpt_data_cleaned = tpt_data[~condition1].copy()
tpt_data_cleaned = tpt_data_cleaned[~condition2]

# Display the cleaned dataset
print(tpt_data_cleaned.shape)

(7258, 34)


<ipython-input-8-ccc75c2ebaa0>:16: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  tpt_data_cleaned = tpt_data_cleaned[~condition2]


## **Removing raws without any author**

In [ ]:
# Remove rows where 'authors' is an empty string
# PE
pe_data_cleaned = pe_data_cleaned.dropna(subset=['authors'])
pe_data_cleaned = pe_data_cleaned[pe_data_cleaned['authors'] != '']

# TPT
tpt_data_cleaned = tpt_data_cleaned.dropna(subset=['author_list'])
tpt_data_cleaned = tpt_data_cleaned[tpt_data_cleaned['author_list'] != '']

In [ ]:
print("PE: ", pe_data_cleaned.shape)
print("TPT: " ,tpt_data_cleaned.shape)

PE:  (6899, 11)
TPT:  (7258, 34)


In [ ]:
# Remove rows where 'fulltext' is an empty string
# PE
pe_data_cleaned = pe_data_cleaned.dropna(subset=['fulltext'])
pe_data_cleaned = pe_data_cleaned[pe_data_cleaned['fulltext'] != '']

# TPT
tpt_data_cleaned = tpt_data_cleaned.dropna(subset=['fulltext'])
tpt_data_cleaned = tpt_data_cleaned[tpt_data_cleaned['fulltext'] != '']

print("PE: ", pe_data_cleaned.shape)
print("TPT: " ,tpt_data_cleaned.shape)

PE:  (6697, 11)
TPT:  (7258, 34)


## **counting words number**

we counted how many rows will be eliminated by setting different thresholds: 300-400-500

In [ ]:
# identifying rows having fulltext less than "number" words

# TPT
# Assuming tpt_data_cleaned is your DataFrame and 'fulltext' contains the text
tpt_data_cleaned['word_count'] = tpt_data_cleaned['fulltext'].apply(lambda x: len(str(x).split()))

# Filter rows where word count is less than [number]
rows_less_than_number = tpt_data_cleaned[tpt_data_cleaned['word_count'] < 300]
rows_less_than_number.shape

# results: 500 --> 55 / 400 --> 29 / 300 --> 17

(17, 35)

In [ ]:
# PE
# Assuming tpt_data_cleaned is your DataFrame and 'fulltext' contains the text
pe_data_cleaned['word_count'] = pe_data_cleaned['fulltext'].apply(lambda x: len(str(x).split()))

# Filter rows where word count is less than [number]
rows_less_than_number = pe_data_cleaned[pe_data_cleaned['word_count'] < 300]

rows_less_than_number.shape

# results: 500 --> 252 / 400 --> 153 / 300 --> 76

(76, 12)

In [ ]:
# Remove rows from pe_data_cleaned and tpt_data_cleanedwhere 'fulltext' is less than 500 words

#PE
pe_data_cleaned = pe_data_cleaned[pe_data_cleaned['word_count'] >= 500]
print("PE: ", pe_data_cleaned.shape)

# TPT
tpt_data_cleaned = tpt_data_cleaned[tpt_data_cleaned['word_count'] >= 500]
print("TPT: ", tpt_data_cleaned.shape)

PE:  (6445, 12)
TPT:  (7203, 35)


# **Final dataset**

**6445** for PE & **7203** for TPT

## Reset index & saving

In [ ]:
# Reset the index of the DataFrame
pe_data_cleaned = pe_data_cleaned.reset_index(drop=True)
tpt_data_cleaned = tpt_data_cleaned.reset_index(drop=True)

# Save the cleaned DataFrame as a pickle file
with open(directory_name + 'physics_education_metadata_with_fulltext_cleaned_v2.pkl', 'wb') as f:
    pickle.dump(pe_data_cleaned, f)

with open(directory_name + 'TPT_metadata_cleaned_fulltext_cleaned_v2.pkl', 'wb') as f:
    pickle.dump(tpt_data_cleaned, f)

In [ ]:
pe_data_cleaned.sample(5)

title  \
4397                     Sliding around a vertical loop   
2944                                       The electron   
4865  Teaching physics by means of cartoons: a quali...   
871                                     Solution: fuses   
1661                     Tabletop divergent-light halos   

                                                authors publication_year  \
4397                                              Cross             2022   
2944                                        D B Gayther             1971   
4865  F Javier Perales-Palacios, José M\nVílchez-Gon...             2002   
871                                       Guy S M Moore             1988   
1661                                   Gislén, Mattsson             2007   

                             doi volume issue fpage lpage  \
4397    10.1088/1361-6552/ac902a     57     6   583     8   
2944   10.1088/0031-9120/6/6/003      6     6   406   412   
4865  10.1088/0031-9120/37/5/306     37     5   400   406   
871   10.1088/0031-9120/23/4/313     23     4   238   238   
1661  10.1088/0031-9120/42/6/003     42     6   579   584   

             pdf_filename                                       zip_filename  \
4397  ped_57_6_065016.pdf  1_0031-9120_57_6_065016_10__1088_1361-6552_ac9...   
2944       pev6i6p406.pdf  1_0031-9120_6_6_406_10__1088_0031-9120_6_6_003...   
4865           pe2506.pdf  1_0031-9120_37_5_400_10__1088_0031-9120_37_5_3...   
871       pev23i4p238.pdf  1_0031-9120_23_4_238_10__1088_0031-9120_23_4_3...   
1661        pe7_6_003.pdf  1_0031-9120_42_6_579_10__1088_0031-9120_42_6_0...   

                                               fulltext  word_count  
4397  PAPER \nPhys. Educ. 57 (2022) 065016 (4pp) iop...        1862  
2944  The atomic particles \nTheelectron \nD B Gayth...        4749  
4865  www.iop.org/journals/physed \nTeaching physics...        3589  
871   Phys Educ 23 11988) Printed In the UK \ntwo gr...         618  
1661  SPECIAL FEATURE: COLOUR \nwww.iop.org/journals...        2469

In [ ]:
tpt_data_cleaned.sample(5)

index       filename  year  \
6923  14575  1.5084920.pdf  2019   
2429   5678  1.2341869.pdf  1985   
6375  13543  1.4908079.pdf  2015   
3205   7195  1.2343389.pdf  1991   
930    2503  1.2350172.pdf  1973   

                                                  title  \
6923  Introducing ROAVEE: An Advanced STEM-Based Pro...   
2429                            Millikan award for 1985   
6375         Making the Most of Your First Day of Class   
3205                        A simple ink‐jet gravimeter   
930                             "Would You Believe...?"   

                               author_list volume issue  processed_len  \
6923             [Hartigan C, Hademenos G]     57     1          15502   
2429                [Meyer JP, Gerhart JB]     23     7           6858   
6375            [Gaffney JDH, Whitaker JT]     53     3          18070   
3205  [Street JG, Fung‐A‐Fat J, Faller JE]     29     7           8301   
930                               [Iona M]     11     9          12477   

         page  page_len  ...  editor  reference-count  is-referenced-by-count  \
6923    17-20         4  ...      []                4                       1   
2429  419-420         2  ...      []                0                       1   
6375  137-139         3  ...      []               20                       5   
3205  468-469         2  ...      []                0                       0   
930   536-539         4  ...      []                0                       1   

          issued                                               link  \
6923   [2019, 1]  http://aapt.scitation.org/doi/pdf/10.1119/1.50...   
2429  [1985, 10]  http://aapt.scitation.org/doi/pdf/10.1119/1.23...   
6375   [2015, 3]  http://aapt.scitation.org/doi/pdf/10.1119/1.49...   
3205  [1991, 10]  http://aapt.scitation.org/doi/pdf/10.1119/1.23...   
930   [1973, 12]  http://aapt.scitation.org/doi/pdf/10.1119/1.23...   

            doi                                           fulltext  \
6923  1.5084920  Introducing ROAVEE: An Advanced STEM-Based Pro...   
2429  1.2341869  Millikan Award for 1985 \nJames Basil Gerhart ...   
6375  1.4908079  Making the Most of Your First Day of Class \nJ...   
3205  1.2343389  APPARATUS \nfor Teaching Physics \nSection Edi...   
930   1.2350172  and Sound," and "Infrared Radiation." His dire...   

                first_n_words  \
6923   INTRODUCING ROAVEE: AN   
2429       MILLIKAN AWARD FOR   
6375          MAKING THE MOST   
3205         A SIMPLE INK‐JET   
930   "WOULD YOU BELIEVE...?"   

                                       cleaned_fulltext word_count  
6923  Introducing ROAVEE: An Advanced STEM-Based Pro...       2342  
2429  Millikan Award for 1985 \nJames Basil Gerhart ...       1144  
6375  Making the Most of Your First Day of Class \nJ...       2742  
3205  APPARATUS \nfor Teaching Physics \nSection Edi...       1384  
930   and Sound," and "Infrared Radiation." His dire...       2075  

[5 rows x 35 columns]